File: sources/academy/webscraping/scraping_basics_javascript/exercises/wta_tennis_players.mjs
import * as cheerio from 'cheerio';

async function download(url) {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }
  const html = await response.text();
  return cheerio.load(html);
}

const listingUrl = 'https://www.wtatennis.com/rankings/singles';
const $listing = await download(listingUrl);
const playerLinks = $listing('.rankings__list .player-row-drawer__link').toArray();

for (const element of playerLinks.slice(0, 5)) {
  const playerUrlRelative = $listing(element).attr('href');
  const playerUrl = new URL(playerUrlRelative, listingUrl).href;
  const $player = await download(playerUrl);

  for (const infoBlock of $player('.profile-bio__info-block').toArray()) {
    const $infoBlock = $player(infoBlock);
    const label = $infoBlock.find('h2').text().trim().toLowerCase();
    if (label === 'birthplace') {
      const birthplace = $infoBlock.find('span').text().trim();
      console.log(`${playerUrl} | ${birthplace}`);
      break;
    }
  }
}

File: sources/academy/webscraping/scraping_basics_python/exercises/wta_tennis_players.py
import httpx
from bs4 import BeautifulSoup
from urllib.parse import urljoin


def download(url: str) -> BeautifulSoup:
    response = httpx.get(url)
    response.raise_for_status()
    return BeautifulSoup(response.text, "html.parser")


listing_url = "https://www.wtatennis.com/rankings/singles"
listing_soup = download(listing_url)
player_links = listing_soup.select(".rankings__list .player-row-drawer__link")

for link in player_links[:5]:
    player_url = urljoin(listing_url, link["href"])
    player_soup = download(player_url)

    for info_block in player_soup.select(".profile-bio__info-block"):
        label_text = info_block.select_one("h2").text.strip()
        if label_text.lower() == "birthplace":
            birthplace = info_block.select_one("span").text.strip()
            print(player_url, "|", birthplace)
